src/OrchardCore/Orchard.Environment.Extensions/ExtensionManager.cs

```diff
@@ -286,72 +282,69 @@ private void EnsureInitialized()
                     loadedExtensions.TryAdd(extension.Id, entry);
                 });
 
-                var loadedFeatures =
-                    new Dictionary<string, FeatureEntry>();
+                var loadedFeatures = new Dictionary<string, FeatureEntry>();
+
+                // Get all valid types from any extension
+                var allTypesByExtension = loadedExtensions.SelectMany(extension =>
+                    extension.Value.ExportedTypes.Where(IsComponentType)
+                    .Select(type => new
+                    {
+                        ExtensionEntry = extension.Value,
+                        Type = type
+                    })).ToArray();
+
+                var typesByFeature = allTypesByExtension
+                    .GroupBy(typeByExtension => GetSourceFeatureNameForType(
+                        typeByExtension.Type, 
+                        typeByExtension.ExtensionEntry.ExtensionInfo.Id))
+                    .ToDictionary(
+                        group => group.Key, 
+                        group => group.Select(typesByExtension => typesByExtension.Type).ToArray());
```

Comment:
I got it, how you grab types by features by grouping from types by extensions, and according to Feature attributes.